### PR TITLE
Add LRN efficient GPU implement.

### PR DESCRIPTION
**Pull request by gongweibao**
Fix https://github.com/PaddlePaddle/Paddle/issues/5066

### CODE REVIEW

File: paddle/operators/lrn_op.cc [hunk 1]
```diff
@@ -19,6 +19,108 @@ namespace operators {
 
 using framework::Tensor;
 
+template <typename T>
+struct LRNFunctor<platform::CPUPlace, T> {
+  void operator()(const framework::ExecutionContext& ctx,
+                  const framework::Tensor* input, framework::Tensor* out,
```
Comment (qingqing01) [inline paddle/operators/lrn_op.cc]: For input arguments: `const framework::Tensor& `

https://google.github.io/styleguide/cppguide.html#Reference_Arguments

File: paddle/operators/lrn_op.cc [hunk 2]
```diff
@@ -19,6 +19,108 @@ namespace operators {
...
+                  framework::Tensor* mid, int N, int C, int H, int W, int n,
+                  T k, T alpha, T beta) {
+    auto x_v = framework::EigenVector<T>::Flatten(*input);
+
+    const int start = -(n - 1) / 2;
+    const int end = start + n;
+
+    auto e_mid = framework::EigenTensor<T, 4>::From(*mid);
+    e_mid.device(ctx.GetEigenDevice<platform::CPUPlace>()) = e_mid.constant(k);
```
Comment (qingqing01) [inline paddle/operators/lrn_op.cc]: For the CPU implementation of Eigen, there is no need to use `.device()`.

```cpp
e_mid.setConstant(k);
```

File: paddle/operators/lrn_op.cc [hunk 3]
```diff
@@ -19,6 +19,108 @@ namespace operators {
...
+          int ch = i + c;
+          if (ch >= 0 && ch < C) {
+            auto s = e_mid.slice(Eigen::array<int, 4>({{m, i, 0, 0}}),
+                                 Eigen::array<int, 4>({{1, 1, H, W}}));
+
+            auto r = e_x.slice(Eigen::array<int, 4>({{m, ch, 0, 0}}),
+                               Eigen::array<int, 4>({{1, 1, H, W}}));
+
+            s.device(ctx.GetEigenDevice<platform::CPUPlace>()) +=
+                alpha * r.square();
```
Comment (qingqing01) [inline paddle/operators/lrn_op.cc]: The same as above:

```cpp
s += alpha * r.square();
```

File: paddle/operators/lrn_op.cc [hunk 4]
```diff
@@ -19,6 +19,108 @@ namespace operators {
...
+          }
+        }
+      }
+    }
+
+    auto out_e = framework::EigenVector<T>::Flatten(*out);
+    out_e.device(ctx.GetEigenDevice<platform::CPUPlace>()) =
+        x_v * e_mid.reshape(Eigen::DSizes<int, 1>(e_mid.size())).pow(-beta);
```
Comment (qingqing01) [inline paddle/operators/lrn_op.cc]: The same as above.

File: paddle/operators/lrn_op.cc [hunk 5]
```diff
@@ -19,6 +19,108 @@ namespace operators {
...
+};
+template struct LRNFunctor<platform::CPUPlace, float>;
+template struct LRNFunctor<platform::CPUPlace, double>;
+
+template <typename T>
+struct LRNGradFunctor<platform::CPUPlace, T> {
+  void operator()(const framework::ExecutionContext& ctx,
+                  const framework::Tensor* x, const framework::Tensor* out,
+                  const framework::Tensor* mid, framework::Tensor* x_g,
+                  const framework::Tensor* out_g, int N, int C, int H, int W,
```
Comment (qingqing01) [inline paddle/operators/lrn_op.cc]: For the input arguments, the same as above comments.

File: paddle/operators/lrn_op.cc [hunk 6]
```diff
@@ -19,6 +19,108 @@ namespace operators {
...
+                                   Eigen::array<int, 4>({{1, 1, H, W}}));
+
+          auto c_mid = e_mid.slice(Eigen::array<int, 4>({{m, ch, 0, 0}}),
+                                   Eigen::array<int, 4>({{1, 1, H, W}}));
+
+          auto c_out_g = e_out_g.slice(Eigen::array<int, 4>({{m, ch, 0, 0}}),
+                                       Eigen::array<int, 4>({{1, 1, H, W}}));
+
+          i_x_g.device(ctx.GetEigenDevice<platform::CPUPlace>()) +=
+              ratio * c_out_g * c_out * i_x / c_mid;
```
Comment (qingqing01) [inline paddle/operators/lrn_op.cc]: 所有`.device()`的地方，如上面评论，感觉可以去掉，也可以保留？

File: paddle/operators/lrn_op.cu [hunk 7]
```diff
@@ -15,8 +15,168 @@
 #define EIGEN_USE_GPU
 #include "paddle/operators/lrn_op.h"
```
Comment (qingqing01) [inline paddle/operators/lrn_op.cu]: Remove `#define EIGEN_USE_GPU` in line 15.

Comment (qingqing01) [reply paddle/operators/lrn_op.cu]: 这里实际不用`#include "paddle/operators/lrn_op.h"`，而是：
` #include "paddle/framework/op_registry.h"`

File: paddle/operators/lrn_op.cu [hunk 8]
```diff
@@ -15,8 +15,168 @@
...
+void CrossMapNormal(const framework::ExecutionContext& ctx, const T* inputs,
+                    T* outputs, T* mid, int N, int C, int H, int W, int n, T k,
+                    T alpha, T beta) {
+  int img_size = N * H * W;
+  int block_size = 1024;
+  int grid_size = (img_size + 1024 - 1) / 1024;
+
+  const auto& stream =
+      reinterpret_cast<const platform::CUDADeviceContext&>(ctx.device_context())
+          .stream();
```
Comment (qingqing01) [inline paddle/operators/lrn_op.cu]: `framework::ExecutionContext`获取cuda stream时可以用接口：

https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/lookup_table_op.cu#L139

File: paddle/operators/lrn_op.cu [hunk 9]
```diff
@@ -15,8 +15,168 @@
...
+  }
+}
+
+template <typename T>
+void CrossMapNormal(const framework::ExecutionContext& ctx, const T* inputs,
+                    T* outputs, T* mid, int N, int C, int H, int W, int n, T k,
+                    T alpha, T beta) {
+  int img_size = N * H * W;
+  int block_size = 1024;
+  int grid_size = (img_size + 1024 - 1) / 1024;
```
Comment (qingqing01) [inline paddle/operators/lrn_op.cu]: 用block_size替代line 69中的1024.
```cpp
int grid_size = (img_size + block_size - 1) / block_size;
```

File: paddle/operators/lrn_op.cu [hunk 10]
```diff
@@ -15,8 +15,168 @@
...
+
+  const auto& stream =
+      reinterpret_cast<const platform::CUDADeviceContext&>(ctx.device_context())
+          .stream();
+  KeCMRNormFillScale<T><<<grid_size, block_size, 0, stream>>>(
+      img_size, inputs, mid, C, H, W, n, k, alpha);
+
+  int input_size = N * H * W * C;
+  block_size = 1024;
+  grid_size = (input_size + 1024 - 1) / 1024;
```
Comment (qingqing01) [inline paddle/operators/lrn_op.cu]: 同上，用block_size替代line 79中的1024.

File: paddle/operators/lrn_op.cu [hunk 11]
```diff
@@ -15,8 +15,168 @@
...
+
+    T accum = 0;
+    int index = 0;
+    while (index < C + post_pad) {
+      if (index < C) {
+        accum += in[index * step] * in[index * step];
+      }
+      if (index >= size) {
+        accum -= in[(index - size) * step] * in[(index - size) * step];
+      }
```
Comment (qingqing01) [inline paddle/operators/lrn_op.cu]: line 41和line 44中，可以利用寄存器先保存global内存中的数据，这样可以避免多次访问globle内存：

```cpp 
       if (index < C) {
         T val = in[index * step];
         accum += val * val;
       }
       if (index >= size) {
         T val = in[index - size) * step];
         accum -= val * val;
       }

```

File: paddle/operators/lrn_op.cu [hunk 12]
```diff
@@ -15,8 +15,168 @@
...
+                        const T* out, const T* mid, T* x_g, const T* out_g,
+                        int N, int C, int H, int W, int n, T alpha, T beta) {
+  int img_size = N * H * W;
+
+  int block_size = 1024;
+  int grid_size = (img_size + 1024 - 1) / 1024;
+
+  const auto& stream =
+      reinterpret_cast<const platform::CUDADeviceContext&>(ctx.device_context())
+          .stream();
```
Comment (qingqing01) [inline paddle/operators/lrn_op.cu]: 同上，参考https://github.com/PaddlePaddle/Paddle/blob/develop/paddle/operators/lookup_table_op.cu#L139

File: paddle/operators/lrn_op.cu [hunk 13]
```diff
@@ -15,8 +15,168 @@
...
+}
+
+template <typename T>
+void CrossMapNormalGrad(const framework::ExecutionContext& ctx, const T* x,
+                        const T* out, const T* mid, T* x_g, const T* out_g,
+                        int N, int C, int H, int W, int n, T alpha, T beta) {
+  int img_size = N * H * W;
+
+  int block_size = 1024;
+  int grid_size = (img_size + 1024 - 1) / 1024;
```
Comment (qingqing01) [inline paddle/operators/lrn_op.cu]: 同上~

File: paddle/operators/lrn_op.cc (hunk 1 above)
Comment (gongweibao) [reply paddle/operators/lrn_op.cc]: Done!
Thanks!

File: paddle/operators/lrn_op.cc (hunk 2 above)
Comment (gongweibao) [reply paddle/operators/lrn_op.cc]: Done!
Thanks!

File: paddle/operators/lrn_op.cc (hunk 3 above)
Comment (gongweibao) [reply paddle/operators/lrn_op.cc]: Done!
Thanks!

File: paddle/operators/lrn_op.cc (hunk 4 above)
Comment (gongweibao) [reply paddle/operators/lrn_op.cc]: Done!
Thanks!

File: paddle/operators/lrn_op.cc (hunk 5 above)
Comment (gongweibao) [reply paddle/operators/lrn_op.cc]: Done!
Thanks!

File: paddle/operators/lrn_op.cc (hunk 6 above)
Comment (gongweibao) [reply paddle/operators/lrn_op.cc]: Done!
Thanks!

File: paddle/operators/lrn_op.cu (hunk 9 above)
Comment (gongweibao) [reply paddle/operators/lrn_op.cu]: Done!
Thanks!

File: paddle/operators/lrn_op.cu (hunk 10 above)
Comment (gongweibao) [reply paddle/operators/lrn_op.cu]: Done!
Thanks!

File: paddle/operators/lrn_op.cu (hunk 13 above)
Comment (gongweibao) [reply paddle/operators/lrn_op.cu]: Done!
Thanks!

File: paddle/operators/lrn_op.cu (hunk 8 above)
Comment (gongweibao) [reply paddle/operators/lrn_op.cu]: Done!
Thanks!

File: paddle/operators/lrn_op.cu (hunk 12 above)
Comment (gongweibao) [reply paddle/operators/lrn_op.cu]: Done!
Thanks!

File: paddle/operators/lrn_op.cu (hunk 7 above)
Comment (gongweibao) [reply paddle/operators/lrn_op.cu]: 下边有一个模板的特化。模板声明在lrn_op.h里边。

**Pull request merged**